File: javascript/word-count/word-count.js
var Words = function () {};

Words.prototype.count = function(word){
  var counts = {};
  var regex = /\s+/;
  var trailingWhitespace = /^\s+|\s+$/g;

  if (trailingWhitespace.test(word)){
    word = word.replace(trailingWhitespace, "");
  }
  
  var words_array = word.replace(regex, " ")
      .split(" ");

  words_array.forEach(function(word) {
    word = word.toLowerCase();
    if(counts[word] === undefined){
      counts[word] = 1;
    } else {
      if(typeof counts[word] === 'function'){
        counts[word] = 0;
      }
      counts[word]++;
    }
  });

  return counts;
}

module.exports = Words;

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> String.replace(~r/[.,\/#!$%\^&\*;:{}=\`~()@]/, "")
    |> String.downcase
    |> String.split([" ", "_"], trim: true)
    |> Enum.reduce(%{},
        fn word, countMap ->
          case Map.fetch(countMap, word) do
            {:ok, count} -> Map.update(countMap, word, count, &(&1 + 1))
            :error -> Map.put(countMap, word, 1)
          end
        end
      )
  end
end

# ["word"] => ${"word" => 0}
